cmd/registry/cmd/upload/bulk/discovery.go

```diff
@@ -155,23 +145,37 @@ func (task *uploadDiscoveryTask) createVersion(ctx context.Context) error {
 	return nil
 }
 
-func (task *uploadDiscoveryTask) createSpec(ctx context.Context) error {
-	if _, err := task.client.GetApiSpec(ctx, &rpc.GetApiSpecRequest{
+func (task *uploadDiscoveryTask) createOrUpdateSpec(ctx context.Context) error {
+	// Use the spec size and hash to avoid unnecessary uploads.
+	if spec, err := task.client.GetApiSpec(ctx, &rpc.GetApiSpecRequest{
 		Name: task.specName(),
-	}); !core.NotFound(err) {
-		return err // Returns nil when spec is found without error.
+	}); err == nil {
+		if int(spec.GetSizeBytes()) == len(task.contents) {
+			hash := hashForBytes(task.contents)
+			if spec.GetHash() == hash {
+				log.Debugf("Matched existing spec %s", task.specName())
+				return nil // this spec is already uploaded
```

Comment:
I think the log message should include a mention that we're skipping this upload _because_ it matches an existing spec. In addition to being more informative, the comment will become unnecessary.

Reply:
OK - updated.